config.org
#+TITLE: config
#+OPTIONS: toc:4 h:4
#+STARTUP: hideblocks
#+PROPERTY: header-args :results silent :tangle yes

* Roadmap
** TODO Encrypted cloud storage of org files
** TODO authsource.gpg or authinfo.gpg
* Notes
** Display crash last time after opening a scratch buffer in clojure, alt tabbing away to zoom
*** Was also running vterm, blue theme
* Ideas and inspiration
** https://sam217pa.github.io/2016/09/02/how-to-build-your-own-spacemacs/
** Sacha Chua's config
https://pages.sachachua.com/.emacs.d/Sacha.html
** Tecosaur's doom config (great ideas here)
https://github.com/tecosaur/emacs-config/blob/master/config.org
** Robin's config
https://github.com/sulami/dotfiles/blob/master/emacs/.emacs/README.org
** Stig's config
https://github.com/stig/dotfiles/blob/trunk/Emacs.org
* Package Configuration
** Clojure Language
*** CIDER (Clojure)
#+begin_src emacs-lisp
;; (after! cider
;;   (cider-auto-test-mode 1))
#+end_src
*** flycheck-clojure
#+begin_src emacs-lisp
;; (use-package! flycheck-clojure
;;   :defer t
;;   :commands (flycheck-clojure-setup)               ;; autoload
;;   :config
;;   (eval-after-load 'flycheck
;;     '(setq flycheck-display-errors-function #'flycheck-pos-tip-error-messages))
;;   (add-hook 'after-init-hook #'global-flycheck-mode))
;; (after! cider (flycheck-clojure-setup))
;; (use-package! flycheck-pos-tip
;;   :defer t
;;   :after flycheck)
#+end_src
*** Clojure REPL
#+begin_src emacs-lisp
  (map! :map cider-repl-mode-map
        :ni "<up>" #'cider-repl-previous-input
        :ni "<down>" #'cider-repl-next-input)
#+end_src
*** Clojure mode
#+begin_src emacs-lisp
(after! clojure-mode
  (setq clojure-indent-style 'align-arguments))
#+end_src
** Company
#+begin_src emacs-lisp
(after! company
  (setq company-idle-delay 0.5
        company-minimum-prefix-length 2)
  (setq company-show-numberse t)
  (add-hook 'evil-normal-state-entry-hook #'company-abort))
#+end_src
** counsel-dash
#+BEGIN_SRC emacs-lisp
(use-package! counsel-dash
  :config
  (add-hook 'emacs-lisp-mode-hook (lambda () (setq-local counsel-dash-docsets '("Emacs Lisp"))))
  (add-hook 'ruby-mode-hook (lambda () (setq-local counsel-dash-docsets '("Ruby")))))
#+END_SRC
** mood-line
#+begin_src emacs-lisp
(use-package! mood-line
  :ensure t
  :config
  (mood-line-mode))
#+end_src
** Elfeed Configuration
#+begin_src emacs-lisp
  (setq rmh-elfeed-org-files (list "~/.doom.d/elfeed.org"))
#+end_src
** Graphviz settings
#+begin_src emacs-lisp
(use-package! graphviz-dot-mode
  :defer t
  :config
  (setq graphviz-dot-indent-width 4))
(use-package! company-graphviz-dot
  :after
  graphviz-dot-mode)
#+end_src
** Info colors
#+begin_src emacs-lisp
(use-package! info-colors
  :commands (info-colors-fontify-node)
  :after
  (add-hook 'Info-selection-hook 'info-colors-fontify-node)
  (add-hook 'Info-mode-hook #'mixed-pitch-mode))
#+end_src
** Ivy
#+begin_src emacs-lisp
(after! ivy
  (setq ivy-read-action-function #'ivy-hydra-read-action)
  (setq ivy-sort-max-size 50000))
#+end_src
** Javascript
#+begin_src emacs-lisp
(use-package! prettier-js
:hook ((js2-mode-hook . prettier-js-mode)
       (web-mode-hook . prettier-js-mode)))
#+end_src
** Very Large Files
#+begin_src emacs-lisp
(use-package! vlf-setup
  :defer-incrementally vlf-tune vlf-base vlf-write vlf-search vlf-occur vlf-follow vlf-ediff vlf)
#+end_src
** Which key
Decrease the idle delay
#+begin_src emacs-lisp
(after! which-key
  (setq which-key-idle-delay 0.5))
#+end_src
** YASnippet
Nested snippets?
#+begin_src emacs-lisp
(setq yas-triggers-in-field t)
#+end_src
* Custom Configuration
** Add an exec path
#+BEGIN_SRC emacs-lisp
(setenv "PATH" (concat (getenv "PATH") ":/usr/local/bin"))
(setq exec-path (append exec-path '("/usr/local/bin")))
#+END_SRC
* Key Configuration
**This is my leader config which overrides some of the other leader configs
in default doom-emacs
#+BEGIN_SRC emacs-lisp
;; (map!
;;       :mnoeivg "C-n" #'next-line
;;       :mnoeivg "C-p" #'previous-line
;; )

(map! :leader
      :desc "M-x" "SPC" #'execute-extended-command
      :desc "Search project" "/"
      (cond ((featurep! :completion ivy)  #'+ivy/project-search)
            ((featurep! :completion helm) #'+helm/project-search))
      (:prefix ("b" . "buffer")
        :desc "ibuffer" "i" #'ibuffer)

      (:prefix ("f")
       :desc "Toggle Treemacs" "t" #'treemacs
        :nv "y" nil)

      (:prefix "o"
        :desc "Elfeed (RSS Feed)"          "e" #'elfeed)

      (:prefix ("p")
        :desc "Find file in project" "f" #'projectile-find-file)

      (:prefix ("r" . "resume")
        :desc "Resume Ivy" "l" #'ivy-resume
        :desc "Resume Avy" "a" #'avy-resume)

      (:prefix ("s")
        :desc "Jump to symbol across buffers" "I" #'imenu-anywhere
        :desc "Search buffer"                 "b" #'swiper
        :desc "Search current directory"      "d"
        (cond ((featurep! :completion ivy)  #'+ivy/project-search-from-cwd)
              ((featurep! :completion helm) #'+helm/project-search-from-cwd))
        :desc "Jump to symbol"                "i" #'imenu
        :desc "Jump to link"                  "l" #'ace-link
        :desc "Look up online"                "o" #'+lookup/online-select
        :desc "Search project"                "p"
        (cond ((featurep! :completion ivy)  #'+ivy/project-search)
              ((featurep! :completion helm) #'+helm/project-search))
        :desc "deadgrep"                      ";" #'deadgrep)

      (:prefix ("w")
       :desc "Edit windows" "e" #'+hydra/window-nav/body)

      (:prefix ("y" . "snippets")
        :desc "New snippet"                "n" #'yas-new-snippet
        :desc "Insert snippet"             "i" #'yas-insert-snippet
        :desc "Jump to mode snippet"       "/" #'yas-visit-snippet-file
        :desc "Jump to snippet"            "s" #'+snippets/find-file
        :desc "Browse snippets"            "S" #'+snippets/browse
        :desc "Reload snippets"            "r" #'yas-reload-all)
      )
(map!
  ;; Easier window navigation
  :n "-"     #'dired-jump
  :n "C-s"   #'counsel-grep-or-swiper
  :n "gb" #'browse-url
  :n "gsr" #'avy-resume
)

(setq mac-option-modifier 'meta)
#+END_SRC
* Editor Configuration
** Indents
#+BEGIN_SRC emacs-lisp
;; (setq-default evil-shift-width 2)
  (setq-default tab-width 2)
#+END_SRC
*** Web Mode
#+begin_src emacs-lisp
  (setq-default web-mode-code-indent-offset 2)
  (setq-default web-mode-markup-indent-offset 2)
  (setq-default web-mode-attr-indent-offset 2)
  (setq-default web-mode-attr-value-indent-offset 2)
#+end_src
*** Typescript
#+begin_src emacs-lisp
  (setq-default typescript-indent-level 2)
#+end_src
*** Ruby
#+begin_src emacs-lisp
(setq-default ruby-indent-level 2)
#+end_src
** Line Numbering
#+BEGIN_SRC emacs-lisp
;; Set line numbers to be relative
;; (setq display-line-numbers 'relative)
;; (setq display-line-numbers-type 'relative)
;; (setq display-line-numbers-current-absolute t)
;; (global-display-line-numbers-mode t)
#+END_SRC
** Theme
#+BEGIN_SRC emacs-lisp
  ;; (if (display-graphic-p) (setq doom-theme 'doom-nord))
(load-theme 'doom-vibrant t)
#+END_SRC
** Font
#+BEGIN_SRC elisp
  (setq doom-font (font-spec :family "Fira Code Retina" :size 16)
        doom-variable-pitch-font (font-spec :family "Fira Code Retina" :size 14))
#+END_SRC
** Menu Bar
Hide the menu bar if it isn't hidden
#+BEGIN_SRC elisp
  ;; (menu-bar-mode -1)
  ;; (toggle-scroll-bar -1)
  ;; (tool-bar-mode -1)
#+END_SRC
** Org Mode
Set org agenda files
#+begin_src emacs-lisp
(setq org-agenda-files (quote ("~/org" "~/.deft")))
#+end_src
** Start in maximized screen
#+BEGIN_SRC elisp
(toggle-frame-maximized)
#+END_SRC
** Start the emacs server
#+begin_src emacs-lisp
(server-start)
#+end_src
** Cursor color
#+begin_src emacs-lisp
(setq evil-emacs-state-cursor '("pink" box))
(setq evil-normal-state-cursor '("pink" box))
(setq evil-visual-state-cursor '("firebrick" box))
(setq evil-insert-state-cursor '("yellow" bar))
(setq evil-replace-state-cursor '("red" bar))
(setq evil-operator-state-cursor '("red" hollow))
#+end_src
* Better defaults
** Simple Settings
Borrowed from Tecosaur's config
#+begin_src emacs-lisp
(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t)                              ; Stretch cursor to the glyph width

(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      truncate-string-ellipsis "…")               ; Unicode ellispis are nicer than "...", and also save /precious/ space
#+end_src
* NOTES
:properties:
:header-args: :tangle no
:end:
** Error message: "Could not find package X in recipe repositories"
*** https://github.com/hlissner/doom-emacs/issues/2802
#+BEGIN_SRC sh
  # Update MELPA
  cd ~/.emacs.d/.local/straight/repos/melpa
  git pull

  # Clear straight's cache
  rm -f ~/.emacs.d/.local/straight/build-cache.el
  doom sync
#+END_SRC

** Compiled Emacs terminal and other stuff:
https://emacsformacosx.com/tips
** If issue with recursive load
*** From https://github.com/bbatsov/prelude/issues/1134
`(setq load-prefer-newer t)`


config.el
;; (after! cider
;;   (cider-auto-test-mode 1))

;; (use-package! flycheck-clojure
;;   :defer t
;;   :commands (flycheck-clojure-setup)               ;; autoload
;;   :config
;;   (eval-after-load 'flycheck
;;     '(setq flycheck-display-errors-function #'flycheck-pos-tip-error-messages))
;;   (add-hook 'after-init-hook #'global-flycheck-mode))
;; (after! cider (flycheck-clojure-setup))
;; (use-package! flycheck-pos-tip
;;   :defer t
;;   :after flycheck)

(map! :map cider-repl-mode-map
      :ni "<up>" #'cider-repl-previous-input
      :ni "<down>" #'cider-repl-next-input)

(after! clojure-mode
  (setq clojure-indent-style 'align-arguments))

(after! company
  (setq company-idle-delay 0.5
        company-minimum-prefix-length 2)
  (setq company-show-numberse t)
  (add-hook 'evil-normal-state-entry-hook #'company-abort))

(use-package! counsel-dash
  :config
  (add-hook 'emacs-lisp-mode-hook (lambda () (setq-local counsel-dash-docsets '("Emacs Lisp"))))
  (add-hook 'ruby-mode-hook (lambda () (setq-local counsel-dash-docsets '("Ruby")))))

(use-package! mood-line
  :ensure t
  :config
  (mood-line-mode))

(setq rmh-elfeed-org-files (list "~/.doom.d/elfeed.org"))

(use-package! graphviz-dot-mode
  :defer t
  :config
  (setq graphviz-dot-indent-width 4))
(use-package! company-graphviz-dot
  :after
  graphviz-dot-mode)

(use-package! info-colors
  :commands (info-colors-fontify-node)
  :after
  (add-hook 'Info-selection-hook 'info-colors-fontify-node)
  (add-hook 'Info-mode-hook #'mixed-pitch-mode))

(after! ivy
  (setq ivy-read-action-function #'ivy-hydra-read-action)
  (setq ivy-sort-max-size 50000))

(use-package! prettier-js
:hook ((js2-mode-hook . prettier-js-mode)
       (web-mode-hook . prettier-js-mode)))

(use-package! vlf-setup
  :defer-incrementally vlf-tune vlf-base vlf-write vlf-search vlf-occur vlf-follow vlf-ediff vlf)

(after! which-key
  (setq which-key-idle-delay 0.5))

(setq yas-triggers-in-field t)

(setenv "PATH" (concat (getenv "PATH") ":/usr/local/bin"))
(setq exec-path (append exec-path '("/usr/local/bin")))

;; (map!
;;       :mnoeivg "C-n" #'next-line
;;       :mnoeivg "C-p" #'previous-line
;; )

(map! :leader
      :desc "M-x" "SPC" #'execute-extended-command
      :desc "Search project" "/"
      (cond ((featurep! :completion ivy)  #'+ivy/project-search)
            ((featurep! :completion helm) #'+helm/project-search))
      (:prefix ("b" . "buffer")
        :desc "ibuffer" "i" #'ibuffer)

      (:prefix ("f")
       :desc "Toggle Treemacs" "t" #'treemacs
        :nv "y" nil)

      (:prefix "o"
        :desc "Elfeed (RSS Feed)"          "e" #'elfeed)

      (:prefix ("p")
        :desc "Find file in project" "f" #'projectile-find-file)

      (:prefix ("r" . "resume")
        :desc "Resume Ivy" "l" #'ivy-resume
        :desc "Resume Avy" "a" #'avy-resume)

      (:prefix ("s")
        :desc "Jump to symbol across buffers" "I" #'imenu-anywhere
        :desc "Search buffer"                 "b" #'swiper
        :desc "Search current directory"      "d"
        (cond ((featurep! :completion ivy)  #'+ivy/project-search-from-cwd)
              ((featurep! :completion helm) #'+helm/project-search-from-cwd))
        :desc "Jump to symbol"                "i" #'imenu
        :desc "Jump to link"                  "l" #'ace-link
        :desc "Look up online"                "o" #'+lookup/online-select
        :desc "Search project"                "p"
        (cond ((featurep! :completion ivy)  #'+ivy/project-search)
              ((featurep! :completion helm) #'+helm/project-search))
        :desc "deadgrep"                      ";" #'deadgrep)

      (:prefix ("w")
       :desc "Edit windows" "e" #'+hydra/window-nav/body)

      (:prefix ("y" . "snippets")
        :desc "New snippet"                "n" #'yas-new-snippet
        :desc "Insert snippet"             "i" #'yas-insert-snippet
        :desc "Jump to mode snippet"       "/" #'yas-visit-snippet-file
        :desc "Jump to snippet"            "s" #'+snippets/find-file
        :desc "Browse snippets"            "S" #'+snippets/browse
        :desc "Reload snippets"            "r" #'yas-reload-all)
      )
(map!
  ;; Easier window navigation
  :n "-"     #'dired-jump
  :n "C-s"   #'counsel-grep-or-swiper
  :n "gb" #'browse-url
  :n "gsr" #'avy-resume
)

(setq mac-option-modifier 'meta)

;; (setq-default evil-shift-width 2)
  (setq-default tab-width 2)

(setq-default web-mode-code-indent-offset 2)
(setq-default web-mode-markup-indent-offset 2)
(setq-default web-mode-attr-indent-offset 2)
(setq-default web-mode-attr-value-indent-offset 2)

(setq-default typescript-indent-level 2)

(setq-default ruby-indent-level 2)

;; Set line numbers to be relative
;; (setq display-line-numbers 'relative)
;; (setq display-line-numbers-type 'relative)
;; (setq display-line-numbers-current-absolute t)
;; (global-display-line-numbers-mode t)

;; (if (display-graphic-p) (setq doom-theme 'doom-nord))
(load-theme 'doom-vibrant t)

(setq org-agenda-files (quote ("~/org" "~/.deft")))

(server-start)

(setq evil-emacs-state-cursor '("pink" box))
(setq evil-normal-state-cursor '("pink" box))
(setq evil-visual-state-cursor '("firebrick" box))
(setq evil-insert-state-cursor '("yellow" bar))
(setq evil-replace-state-cursor '("red" bar))
(setq evil-operator-state-cursor '("red" hollow))

(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t)                              ; Stretch cursor to the glyph width

(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      truncate-string-ellipsis "…")               ; Unicode ellispis are nicer than "...", and also save /precious/ space

(setq doom-font (font-spec :family "Fira Code Retina" :size 16)
      doom-variable-pitch-font (font-spec :family "Fira Code Retina" :size 14))

;; (menu-bar-mode -1)
;; (toggle-scroll-bar -1)
;; (tool-bar-mode -1)

(toggle-frame-maximized)
